### ISSUE TITLE
Remove -H:+AllowIncompleteClasspath and --report-unsupported-elements-at-runtime 

### ISSUE THREAD

**Issue by ppalaga** [GoogleCloudPlatform/google-cloud-graalvm-support#141]
These are options suitable for debugging but nor for running production code. For `--report-unsupported-elements-at-runtime`, see e.g. https://github.com/oracle/graal/issues/1725#issuecomment-547570158 

Related  https://github.com/apache/camel-quarkus/issues/2692 , https://github.com/quarkiverse/quarkus-google-cloud-services/issues/119

**Comment by loicmathieu** [comment]
@dzou can you have a look at this one ?

We discover this flag being set on the library while investiguating a graalvm native issue on recent version of Netty (due to Brotli compresson), and it's bad practice as this is a debugging flag that should be used at client side not at library side.

**Comment by elefeint** [comment]
@ppalaga Thanks for catching this! We'll remove the development properties.

**Comment by dzou** [comment]
Thanks for reporting this, I will try to resolve this.

Unfortunately we are blocked on fixing this immediately; due to an issue in the `grpc-netty-shaded` dependency (See: https://github.com/grpc/grpc-java/issues/7540). Basically this dependency does not shade the native-image config inherited from netty correctly and the compiler will not recognize the classes referenced. This is why we had to add `--report-unsupported-elements-at-runtime` originally.

I'll try to contribute the fix upstream and see if we can resolve this.

**Comment by dzou** [comment]
Just an update, finally got around to fixing this.

We were able to submit a patch in https://github.com/grpc/grpc-java/pull/8258 upstream, waited for a new release, and #167 will fully remove the flags.

**Comment by loicmathieu** [comment]
@dzou great, thanks for the heads up

**Comment by ppalaga** [comment]
Perfect, thanks!

**Comment by dzou** [comment]
@loicmathieu @ppalaga -- Sorry guys, I was a little early in my excitement reporting the news. I was actually only able to remove `--report-unsupported-elements-at-runtime` however unable to remove `allow-incomplete-classpath`.

I'm basically hitting this class of error and not sure how to solve it without `--allow-incomplete-classpath`. 

```
Error: com.oracle.graal.pointsto.constraints.UnresolvedElementException:
Discovered unresolved method during parsing: io.grpc.netty.shaded.io.netty.handler.ssl.JettyAlpnSslEngine$ClientEngine.<init>(javax.net.ssl.SSLEngine, io.grpc.netty.shaded.io.netty.handler.ssl.JdkApplicationProtocolNegotiator).
To diagnose the issue you can use the --allow-incomplete-classpath option. The missing method is then reported at run time when it is accessed the first time.
```

I've studied Quarkus's solution for netty which might be similar to what we wish to do. But we don't have the resources for our project to maintain a set of complex GraalVM substitutions; so I'm unable to move in that direction if that is what it takes. But curious to hear if you may know what we could try.

**Comment by ppalaga** [comment]
@dzou hard to say how much the class `JettyAlpnSslEngine$ClientEngine` that refers to unavailable methods is wired to the rest of your artifacts. If it is actually not necessary, you may try to remove it (and its dependents) from the `grpc-netty-shaded` artifact (using Maven Shade plugin or which ever other tool you use to produce `grpc-netty-shaded`).

**Comment by ppalaga** [comment]
If removing is too complicated (and if the stuff done in `io.grpc.netty.shaded.io.netty.handler.ssl.JettyAlpnSslEngine$ClientEngine.<init>(javax.net.ssl.SSLEngine, io.grpc.netty.shaded.io.netty.handler.ssl.JdkApplicationProtocolNegotiator)` is not required) you could try to replace the `ClientEngine.<init>(javax.net.ssl.SSLEngine, io.grpc.netty.shaded.io.netty.handler.ssl.JdkApplicationProtocolNegotiator)` constructor with an empty stub.

**Comment by dzou** [comment]
Gotcha, thanks for the tips. Will give it another attempt.

**Comment by dzou** [comment]
Hey I revisited this, and we were able to successfully remove `--report-unsupported-elements-at-runtime` in https://github.com/GoogleCloudPlatform/google-cloud-graalvm-support/pull/167

I tried for a few hours to remove `-H:+AllowIncompleteClasspath` but I wasn't successful and still ended up requiring a lot of substitution classes. I could try further, but long term I don't think we have the ability to maintain a large set of substitutions as underlying code changes. I'm planning on just keeping `-H:+AllowIncompleteClasspath` for now.

But I think `--report-unsupported-elements-at-runtime` is the main culprit in all of this, hopefully removing this is most of the problem. I will close this issue for now since we don't plan on pursuing this further, but let me know if new problems with `--allow-incomplete-classpath` emerge.
